### PR TITLE
Rename `fitting_problem` to `ml_problem`

### LINKED ISSUE #42: Add 'ml_' prefix to the main functions
Currently the package has three main functions `fitting_problem()`, `ml_fit()`, `ml_replicate()`. Since we will adopt a simpler name for the package, #39, should we rename `fitting_problem()` to `ml_problem()` for consistency?

### LINKED ISSUE #46: How to generate the RDS files in extdata?
@krlmlr we need to regenerate all toy examples for #42. Could you please help me understand how did you generate them? I can't figure out where to begin. For example, `Joint-Grouped` and `Joint` which script did you use to generate them? 

Thanks!

### LINKED ISSUE #48: Rename as.flat_ml_fit_problem to as_flat_ml_fit_problem
to get rid of the dot.

### PR DESCRIPTION
**Pull request by asiripanich**
This closes #42, closes #46, and closes #48.

### CODE REVIEW

File: R/ml_problem.R [hunk 1]
```diff
@@ -94,27 +112,29 @@ fitting_problem <- function(ref_sample,
                             field_names,
                             individual_controls, group_controls,
                             prior_weights = NULL) {
-  new_fitting_problem(
-    list(
-      refSample = ref_sample, controls = controls, fieldNames = field_names,
-      priorWeights = prior_weights
-    )
+  .Deprecated("ml_problem")
```
Comment (krlmlr) [inline R/ml_problem.R:115]:
```suggestion
  deprecate_soft("mlfit::fitting_problem()", "mlfit::ml_problem()")
```

with a suitable import from the lifecycle package.

File: R/ml_problem.R [hunk 2]
```diff
@@ -86,6 +86,24 @@
...
+  new_ml_problem(
+    list(
+      refSample = ref_sample, controls = controls, fieldNames = field_names,
+      priorWeights = prior_weights
+    )
+  )
+}
+
+#' @export
+#' @rdname ml_problem
```
Comment (krlmlr) [inline R/ml_problem.R:106]:
```suggestion
#' Create a fitting problem
#'
#' Soft-deprecated, new code should use [ml_problem()].
#' @export
#' @keywords internal
```

**Review by krlmlr — COMMENTED** [review]
Looks good, thanks! How did you recreate the `.rds` files?

**Comment by asiripanich** [comment]
> 
> 
> Looks good, thanks! How did you recreate the `.rds` files?

This is what I did:
```r
#' this script converts the toy examples as `fitting_problem` objects
#' to `ml_problem` objects.

toy_example_filedirs <-
  list.files(here::here("inst", "extdata"), full.names = TRUE)

toy_fitting_problems <-
  lapply(toy_example_filedirs, readRDS)

make_ml_problem <- function(x) {
  stopifnot(is(x, "fitting_problem"))
  class(x) <- "ml_problem"
  x
}

toy_ml_problems <-
  lapply(toy_fitting_problems, make_ml_problem)

mapply(function(filedir, toy_example) {
  saveRDS(toy_example, filedir, version = 2)
}, toy_example_filedirs, toy_ml_problems)
```
Would be good if you can document how you created them in the first place in #46. :)

File: R/ml_problem.R (hunk 1 above)
Comment (asiripanich) [reply R/ml_problem.R:115]: shouldn't this be?

```r
deprecate_soft("0.4.0", "mlfit::fitting_problem()", details = "Please use `mlfit::ml_problem()` instead.")
```

File: R/ml_problem.R (hunk 2 above)
Comment (asiripanich) [reply R/ml_problem.R:106]: Will put them all in one commit.

File: R/ml_problem.R (hunk 1 above)
Comment (krlmlr) [reply R/ml_problem.R:115]: Yeah, `deprecate_soft("0.4.0", "mlfit::fitting_problem()", "mlfit::ml_problem()")` should be enough.

**Comment by asiripanich** [comment]
Whoops! that last commit should be `#48`.

**Comment by asiripanich** [comment]
@krlmlr a gentle reminder... :)

**Pull request merged**